{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def ReadFromFASTA(filename, Discriptor = False):\n",
    "    \"\"\"Reads from fasta file then outputs only the sequence as an unbroken string\n",
    "        discriptor: tells funciton wheather or not to includ the discriptor \"\"\"\n",
    "    f = open(filename, \"r\")\n",
    "    lines = f.readlines()\n",
    "    f.close()\n",
    "    discriptor = lines[0]\n",
    "    lines  = lines[1:]\n",
    "    \n",
    "    stringSeq = \"\".join(lines)\n",
    "    stringSeq = stringSeq.replace(\"\\n\",\"\")\n",
    "    if Discriptor:\n",
    "        return((discriptor,stringSeq))\n",
    "    else:\n",
    "        return(stringSeq)\n",
    "\n",
    "\n",
    "def WrightSequenceToFASTA(filename, Description = \"\", Sequence = \"\", lineSize = 60):\n",
    "    \"\"\"Prints a description and sequence to a file\n",
    "    filename: String, Name of the File\n",
    "    Desciption: String, discription of the file, will have \\n added to it before it is printed \n",
    "    Sequence: String or list of strings, the body of what will be written\n",
    "    lineSize: number of charicters on each line, i.e number of charicters befor it puts an \\n. Only applicable if Sequence is a string\n",
    "    \n",
    "    \"\"\"\n",
    "    f = open(filename, \"w\")\n",
    "    if Description != \"\":\n",
    "        f.write(Description + \"\\n\")\n",
    "    \n",
    "    if type(Sequence) == type(\"\"):\n",
    "        if not lineSize:\n",
    "            f.write(Sequence)\n",
    "        else:\n",
    "            number_of_lines = len(Sequence) // lineSize + 1\n",
    "            line_begin = 0\n",
    "            line_end = lineSize\n",
    "            for i in range(number_of_lines):\n",
    "                line_begin = i * lineSize\n",
    "                line_end = (i+1)*lineSize\n",
    "                f.write(Sequence[line_begin:line_end] + \"\\n\")\n",
    "            #if line_begin < len(Sequence):\n",
    "            #    f.write(Sequence[line_begin:len(Sequence)] + \"\\n\")\n",
    "    else:\n",
    "        for line in Sequence:\n",
    "            f.write(line + \"\\n\")\n",
    "    f.close()\n",
    "\n",
    "def DNAtoRNA(s, upper = True, outputUpper = True, reverse = False):\n",
    "    \"\"\" Converts DNA sequence to RNA sequence\n",
    "    s: DNA sequence to be converted\n",
    "    upper: tells function if the values it will receve will be upper case\n",
    "    outputUpper: tells function if the values it returns should be upper case\n",
    "    reverse: tells function to reverse the output\n",
    "    \"\"\"\n",
    "    dic = {\"A\":\"A\", \"T\":\"U\", \"G\":\"G\", \"C\":\"C\"}\n",
    "    oupt = Conversion(dic, s, upper = upper, outputUpper = upper)\n",
    "\n",
    "    if reverse:\n",
    "        return(oupt[::-1])\n",
    "    else:\n",
    "        return(oupt)\n",
    "\n",
    "\n",
    "def RNAtoDNA(s, upper = True, outputUpper = True, reverse = False):\n",
    "    \"\"\" Converts RNA sequence to DNA sequence\n",
    "    s: DNA sequence to be converted\n",
    "    upper: tells function if the values it will receve will be upper case\n",
    "    outputUpper: tells function if the values it returns should be upper case\n",
    "    reverse: tells function to reverse the output\n",
    "    \"\"\"\n",
    "    dic = {\"A\":\"A\", \"U\":\"T\", \"G\":\"G\", \"C\":\"C\"}\n",
    "    oupt = Conversion(dic, s, upper = upper, outputUpper = upper)\n",
    "\n",
    "    if reverse:\n",
    "        return(oupt[::-1])\n",
    "    else:\n",
    "        return(oupt)\n",
    "\n",
    "\n",
    "def DNAtoDNA(s, upper = True, outputUpper = True, reverse = False):\n",
    "    \"\"\" Converts DNA sequence its complement strand \n",
    "    s: DNA sequence to be converted\n",
    "    upper: tells function if the values it will receve will be upper case\n",
    "    outputUpper: tells function if the values it returns should be upper case\n",
    "    reverse: tells function to reverse the output\n",
    "    \"\"\"\n",
    "    dic = {\"A\":\"T\", \"T\":\"A\", \"C\":\"G\", \"G\":\"C\"}\n",
    "    oupt = Conversion(dic, s, upper = upper, outputUpper = upper)\n",
    "\n",
    "    if reverse:\n",
    "        return(oupt[::-1])\n",
    "    else:\n",
    "        return(oupt)\n",
    "\n",
    "def RNAtoRNA(s, upper = True, outputUpper = True, reverse = False):\n",
    "    \"\"\" Converts RNA sequence its complement strand \n",
    "    s: RNA sequence to be converted\n",
    "    upper: tells function if the values it will receve will be upper case\n",
    "    outputUpper: tells function if the values it returns should be upper case\n",
    "    reverse: tells function to reverse the output\n",
    "    \"\"\"\n",
    "    dic = {\"A\":\"U\", \"U\":\"A\", \"C\":\"G\", \"G\":\"C\"}\n",
    "    oupt = Conversion(dic, s, upper = upper, outputUpper = upper)\n",
    "\n",
    "    if reverse:\n",
    "        return(oupt[::-1])\n",
    "    else:\n",
    "        return(oupt)\n",
    "\n",
    "\n",
    "\n",
    "def ReverseStrint(s):\n",
    "    return(s[::-1])\n",
    "\n",
    "\n",
    "\n",
    "def Conversion(dic, s, upper = True, outputUpper = True):\n",
    "    \"\"\"maps a string into a new string using a dictinary\n",
    "    dic: mapping dicionary \n",
    "    s: string to be converted\n",
    "    upper: tells function if the values it will receve will be upper case\n",
    "    outputUpper: tells function if the values it returns should be upper case\n",
    "    \"\"\"\n",
    "    if upper:\n",
    "        s = s.upper()\n",
    "    oupt = \"\"\n",
    "    for i in s:\n",
    "        oupt += dic[i]\n",
    "    if not outputUpper:\n",
    "        oupt = oupt.lower()\n",
    "    return(oupt)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def DNAtoAA(s,upper = True, joiner = \"\", outputUpper = True):\n",
    "        \"\"\"Turns DNA sequnce into the Ameno Acid sequence\n",
    "        s: input sequence\n",
    "        upper: tells function wheather input string is in upper case or lowercase\n",
    "        joiner: tells the function what string it should use to join the ameno acids \n",
    "        outputUpper: Tells the funciton if the output should be capitalised\"\"\"\n",
    "        if not upper:\n",
    "            s = s.upper()\n",
    "        dic =  {\n",
    "                'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "                'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "                'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "                'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "                'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "                'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "                'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "                'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "                'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "                'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "                'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "                'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "                'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "                'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "                'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n",
    "                'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W'}\n",
    "        oupt = []\n",
    "        i1 = 0 \n",
    "        while i1+3 <= len(s):\n",
    "            key = s[i1:i1+3]\n",
    "            AA =    dic[key]   \n",
    "            if not outputUpper:\n",
    "                aa = AA[0].lower()\n",
    "                AA = aa+AA[1:]\n",
    "            oupt.append(AA)\n",
    "            i1 = i1 + 3\n",
    "            \n",
    "        oupt = joiner.join(oupt)\n",
    "        return(oupt)\n",
    "    \n",
    "    \n",
    "import re\n",
    "def find_reading_frames(DNA, MinCodons = 100, StartCodons = [\"ATG\"], StopCodons = [\"TAG\",\"TAA\",\"TGA\"]):\n",
    "    \"\"\"\n",
    "    DNA: DNA sequence we are looking at \n",
    "    MinCodons: Minimum number of codons in the sequence\n",
    "    StartCodons:  list of strings, start codons\n",
    "    StopCodons: list of strings, stop codons\n",
    "    \"\"\"\n",
    "    start = \"(\" + \"|\".join(StartCodons)+\")\"\n",
    "    stop = \"|\".join(StopCodons)\n",
    "    patturn =  start + \"((?!\" + stop + \")...){\" + str(MinCodons) + \",}(\"+stop+\")\"\n",
    "    p = raw_s = r'{}'.format(patturn)\n",
    "    pattern2 = re.compile(p)\n",
    "    orf_list2 = re.finditer(pattern2,DNA)\n",
    "    oupt = []\n",
    "    for each in orf_list2:\n",
    "        dic = {\"Start Index\":each.start(), \"Start Coord\":each.start() + 1,  \"End Index\":each.end(), \"End Coord\":each.end(), \"Sequence\":each.group()}\n",
    "        oupt.append(dic)\n",
    "    return(oupt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA to RNA: CAUGCC\n",
      "RNA to DNA: CATGCC\n",
      "DNA to DNA: GTACGG\n",
      "RNA to RNA: GUACGG\n",
      "MG*LAS*LIDAS*SI*LASYVAS*LASYYVSQSVEYCS\n"
     ]
    }
   ],
   "source": [
    "RNA = \"CAUGCC\"\n",
    "DNA = \"CATGCC\"\n",
    "\n",
    "print(\"DNA to RNA: \" + DNAtoRNA(DNA))\n",
    "print(\"RNA to DNA: \" + RNAtoDNA(RNA))\n",
    "print(\"DNA to DNA: \" + DNAtoDNA(DNA))\n",
    "print(\"RNA to RNA: \" + RNAtoRNA(RNA))\n",
    "\n",
    "\n",
    "Sequence = \"ATGGGCTAGCTAGCTAGCTAGCTGATCGATGCTAGCTGATCGATCTAGCTAGCTAGCTACGTAGCTAGCTAGCTAGCTAGCTACTACGTCAGTCAGTCAGTCGAGTACTGCAGT\"\n",
    "print(DNAtoAA(Sequence)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUGCGUACUGGUAAUGCAAACUAA\n",
      "MRTGNAN*\n",
      "TTAGTTTGCATTACCAGTACGCAT\n"
     ]
    }
   ],
   "source": [
    "DNA = \"ATGCGTACTGGTAATGCAAACTAA\"\n",
    "RNA = DNAtoRNA(DNA)\n",
    "print(RNA)\n",
    "AA = DNAtoAA(DNA)\n",
    "print(AA)\n",
    "DNA_compt = DNAtoDNA(DNA, reverse = True)\n",
    "print(DNA_compt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7 (default, Sep 16 2021, 13:09:58) \n[GCC 7.5.0]"
  },
  "vscode": {
   "interpreter": {
    "hash": "08b152a43b990657280b5bd1f6b4d6b9ebb44a72f30ddd67cde5a313bf6c538b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
